src/Cake.GitHub/Releases/GitHubReleaseCreator.cs

```
@@ -0,0 +1,216 @@
+                {
+                    _cakeLog.Debug($"Uploading asset '{filePath}'");
+                    using var stream = _fileSystem.GetFile(filePath).OpenRead();
+                    var assetUpload = new ReleaseAssetUpload()
+                    {
+                        FileName = filePath.GetFilename().ToString(),
+                        ContentType = "application/octet-stream",
+                        RawData = stream
+                    };
+                    var asset = await _githubClient.Repository.Release.UploadAsset(createdRelease, assetUpload);
+                    result.Add(GitHubReleaseAsset.FromReleaseAsset(asset));
+                }
+            }
+
+            _cakeLog.Verbose(settings.AssetsOrEmpty.Count > 0
+                ? $"Successfully created new GitHub Release and uploaded '{settings.AssetsOrEmpty.Count}' assets"
+                : "Successfully created new GitHub Release");
+
+            return result;
+        }
+
+        private async Task<Release?> TryGetReleaseAsync(string owner, string repository, string tagName)
+        {
+            try
+            {
+                var release = await _githubClient.Repository.Release.Get(owner, repository, tagName);
+                _cakeLog.Verbose($"Found existing release for tag '{tagName}'");
+                return release;
+            }
+            catch (NotFoundException)
+            {
+                // in case a release is a draft release, the tag has not been created yet and the release cannot be found via Get()
+                // to retrieve it, get all releases and search for the tag name                
+                var allReleases = await _githubClient.Repository.Release.GetAll(owner, repository);
+                var matchingReleases = allReleases.Where(x => StringComparer.Ordinal.Equals(x.TagName, tagName)).ToArray();
+
+                Release? release;
+                if(matchingReleases.Length == 0)
```

can be replaced with SingleOrDefault()?

---

I intentionally did not use `SingleOrDefault()` to get a better exception message rather than the generic message given by `SingleOrDefault()` and catching the exception just to rethrow it seemed wrong as well

---

Makes sense, thanks for explaining.